src/signTxCatalystRegistration.c

```diff
@@ -0,0 +1,576 @@
+
+}
+
+static inline void CHECK_STATE(sign_tx_catalyst_registration_state_t expected)
+{
+	TRACE("Catalyst voting registration submachine state: current %d, expected %d", subctx->state, expected);
+	VALIDATE(subctx->state == expected, ERR_INVALID_STATE);
+}
+
+static inline void advanceState()
+{
+	TRACE("Advancing Catalyst registration state from: %d", subctx->state);
+
+	switch (subctx->state) {
+
+	case STATE_CATALYST_REGISTRATION_VOTING_KEY:
+		subctx->state = STATE_CATALYST_REGISTRATION_STAKING_KEY;
+		break;
+
+	case STATE_CATALYST_REGISTRATION_STAKING_KEY:
+		subctx->state = STATE_CATALYST_REGISTRATION_VOTING_REWARDS_ADDRESS;
+		break;
+
+	case STATE_CATALYST_REGISTRATION_VOTING_REWARDS_ADDRESS:
+		subctx->state = STATE_CATALYST_REGISTRATION_NONCE;
+		break;
+
+	case STATE_CATALYST_REGISTRATION_NONCE:
+		subctx->state = STATE_CATALYST_REGISTRATION_CONFIRM;
+		break;
+
+	case STATE_CATALYST_REGISTRATION_CONFIRM:
+		subctx->state = STATE_CATALYST_REGISTRATION_FINISHED;
+		break;
+
+	default:
+		ASSERT(false);
+	}
+
+	TRACE("Advancing Catalyst registration state to: %d", subctx->state);
+}
+
+// ============================== VOTING KEY ==============================
+
+enum {
+	HANDLE_VOTING_KEY_STEP_DISPLAY = 8200,
+	HANDLE_VOTING_KEY_STEP_RESPOND,
+	HANDLE_VOTING_KEY_STEP_INVALID,
+};
+
+static void signTxCatalystRegistration_handleVotingKey_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = signTxCatalystRegistration_handleVotingKey_ui_runStep;
+
+	UI_STEP_BEGIN(subctx->ui_step);
+
+	UI_STEP(HANDLE_VOTING_KEY_STEP_DISPLAY) {
+		STATIC_ASSERT(SIZEOF(subctx->stateData.votingPubKey) == CATALYST_VOTING_PUBLIC_KEY_LENGTH, "wrong voting public key size");
+
+		// Jormungandr public key, hence the "ed25519_pk" prefix
+		// https://github.com/input-output-hk/jormungandr/blob/a057af27493d823be02480bb20258c25ff979e2a/jormungandr-lib/src/crypto/key.rs#L126
+		ui_displayBech32Screen(
+		        "Voting public key",
+		        "ed25519_pk",
+		        subctx->stateData.votingPubKey, CATALYST_VOTING_PUBLIC_KEY_LENGTH,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_VOTING_KEY_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+		advanceState();
+	}
+	UI_STEP_END(HANDLE_VOTING_KEY_STEP_INVALID);
+}
+
+__noinline_due_to_stack__
+static void signTxCatalystRegistration_handleVotingKeyAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		CHECK_STATE(STATE_CATALYST_REGISTRATION_VOTING_KEY);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		explicit_bzero(&subctx->stateData, SIZEOF(subctx->stateData));
+	}
+	{
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+
+		VALIDATE(wireDataSize == SIZEOF(subctx->stateData.votingPubKey), ERR_INVALID_DATA);
+
+		{
+			STATIC_ASSERT(SIZEOF(subctx->stateData.votingPubKey) == CATALYST_VOTING_PUBLIC_KEY_LENGTH, "wrong voting public key size");
+			os_memmove(subctx->stateData.votingPubKey, wireDataBuffer, CATALYST_VOTING_PUBLIC_KEY_LENGTH);
```

Comment:
Using `view_memmove` is preferable IMHO. Might be worth creating an issue and fixing it consistently across all the codebase in a separate PR.